### PR TITLE
Fixes #20821 - Separate katello-* utilities into classes

### PR DESCRIPTION
**Pull request by johnpmitsch**
This moves katello-backup, restore, and hostname change into classes in lib/ that will be called by the actual scripts. This allows us to have more flexibility in changing terminology in backports and also makes the script functionality more testable and compartmentalized.

TO-DO:

- [x] update spec files

### CODE REVIEW

**Comment by johnpmitsch** [comment]
To test, spin up a 3.4 box and run:
```
mkdir /usr/share/katello
cd ~
git clone https://github.com/Katello/katello-packaging
cd katello-packaging/
git remote add fork https://github.com/johnpmitsch/katello-packaging
git fetch --all
git checkout fork/katello_script_refactor
ln -s /root/katello-packaging/katello/lib/restore.rb /usr/share/katello/restore.rb
ln -s /root/katello-packaging/katello/lib/backup.rb /usr/share/katello/backup.rb
ln -s /root/katello-packaging/katello/lib/hostname-change.rb /usr/share/katello/hostname-change.rb
ln -s /root/katello-packaging/katello/lib/script-helper.rb /usr/share/katello/script-helper.rb
cd ~
ln -s /root/katello-packaging/katello/katello-change-hostname katello-change-hostname
ln -s /root/katello-packaging/katello/katello-restore katello-restore
ln -s /root/katello-packaging/katello/katello-backup katello-backup
chmod a+x katello-change-hostname 
chmod a+x katello-restore
chmod a+x katello-backup
```
All backup/restore/change-hostname functionality should be the same as usual

File: katello/lib/backup.rb
Comment (sean797) [inline katello/lib/backup.rb]: I know you haven't changed this but doesn't qpidd have to stop to get a consistent backup of `/var/lib/qpidd` ?

**Review by sean797 — COMMENTED** [review]
Haven't tested this but just looking over it I think changes to the `.spec` file will also be need

File: katello/lib/backup.rb
Comment (johnpmitsch) [reply katello/lib/backup.rb]: Thanks @sean797, I marked it WIP as I do still have that as a change to make, but wanted to get eyes on this early. 

I'm not sure about qpidd  stopping to get an accurate backup, maybe @jlsherrill or @ehelms can comment

**Comment by jturel** [comment]
Just starting my own review but I'm curious if you see any value in one or more of these options:

- Putting the Backup script into module => Katello::Backup.new
- Renaming Backup to BackupCreator which I think makes the intent of the class more clear

Ditto on those suggestions for the other scripts you adjusted.

**Comment by johnpmitsch** [comment]
@jturel I am for having a namespace to make things clear, but one thing I do want to avoid is having a useless subdirectory, i.e. `katello/lib/katello/backup.rb` vs `katello/lib/backup.rb`. I think I can add the modules without the subdirectory (I don't think this breaks conventions or best practices), I'll do that.

Let me think about the names a little more, one idea I have is this:
```
Tools::Backup
Tools::Restore
Tools::HostnameChange
```
or 
```
Utilities::Backup
Utilities::Restore
Utilities::HostnameChange
```
That would signify that these are tools and not actual backups, etc.
I'm not sure if it needs a specific Katello declaration (since we aren't in a plugin situation like we are in actual Katello repo), but open to it if others agree.
\</bikeshedding>

File: katello/katello-backup
Comment (ehelms) [inline katello/katello-backup]: Utilities is a bit of a broad namespace, I do think you'd be better off with something Katello or Foreman specific.

File: katello/katello.spec
Comment (ehelms) [inline katello/katello.spec]: All of these new files will need to be included as Source at the top of the spec.

File: katello/utilities/helper.rb
Comment (ehelms) [inline katello/utilities/helper.rb]: This helper script will make some parts of re-using these scripts more difficult. Or rather, not simply straightforward.

File: katello/katello.spec
Comment (ehelms) [reply katello/katello.spec]: Instead of symlinks, you can see down below, starting at line 98, you can just have them installed to the location you expect them at.

Comment (ehelms) [reply katello/katello.spec]: Take a look at line 135 -- you'll need to include all of these so they get laid down on disk.

File: katello/katello-backup
Comment (johnpmitsch) [reply katello/katello-backup]: KatelloUtilities seems like it would be pretty explicit about what it contains. I'll switch to that unless there are objections

**Comment by jturel** [comment]
I was able to test each of the scripts and see that they are working as intended! I'll accept once the module is renamed. Deferring to Eric on the spec file changes (though, they make sense).

**Comment by ehelms** [comment]
@johnpmitsch What's the plan here? I see a lot of fixes going into some of these scripts thus causing this to need a rebase and update. Which will lead this to need to be re-tested.

**Comment by johnpmitsch** [comment]
@ehelms I have to figure that out, its probably going to be that I manually apply all of the recent commits to the katello-change hostname class

**Comment by johnpmitsch** [comment]
I updated this with the latest katello-change-hostname commits and renamed the `Utitlies` class to `KatelloUtilities`. 
@evgeni  Can you make sure all of the latest commits to katello-change-hostname are found in here? I manually applied them since I couldn't rebase properly with the refactoring.
@ehelms Can you check the spec file?
@jturel Can you retest? Here is what an updated version of what I use to test
```
yum install -y git vim
mkdir /usr/share/katello
cd ~
git clone https://github.com/Katello/katello-packaging
cd katello-packaging/
git remote add fork https://github.com/johnpmitsch/katello-packaging
git fetch --all
git checkout fork/katello_script_refactor
ln -s /root/katello-packaging/katello/utilities/restore.rb /usr/share/katello/restore.rb
ln -s /root/katello-packaging/katello/utilities/backup.rb /usr/share/katello/backup.rb
ln -s /root/katello-packaging/katello/utilities/hostname-change.rb /usr/share/katello/hostname-change.rb
ln -s /root/katello-packaging/katello/utilities/helper.rb /usr/share/katello/helper.rb
cd ~
ln -s /root/katello-packaging/katello/katello-change-hostname katello-change-hostname
ln -s /root/katello-packaging/katello/katello-restore katello-restore
ln -s /root/katello-packaging/katello/katello-backup katello-backup
chmod a+x katello-change-hostname 
chmod a+x katello-restore
chmod a+x katello-backup
```

**Comment by jturel** [comment]
I retested all scripts and they are working well.

File: katello/katello.spec
Comment (mbacovsky) [inline katello/katello.spec]: @johnpmitsch  I'm afraid it is not possible to have resources in subdirectories [1]. It needs to go to the root dir or in a tarball. 

What were the reasons originally to put these scripts here and not in the katello repo?
Was there any agreement on plan to move these utilities to `foreman-maintain`?

[1] http://rikers.org/rpmbook/node68.html#SECTION03526100000000000000

Comment (johnpmitsch) [reply katello/katello.spec]: @mbacovsky The main reason of this refactoring is to make it easy to change our naming conventions for backporting these scripts to satellite. But this is also going to make the scripts more testable and is the first step to moving them to foreman-maintain.

I'm not sure what do do about the spec file and handling the subdirectory, if anyone who is more experienced with rpm spec files can help me out, that would be appreciated

**Comment by johnpmitsch** [comment]
After some off-thread IRC conversations, I updated the scripts to live in the `katello/` dir, everything. I'm currently re-testing the scripts and so far, so good.

@jturel apologies, but do you mind retesting? You will just have to update the links in the setup script that I put in a previous comment. (thanks for being patient)

@ehelms I'm not sure what is going wrong with the build in jenkins, but I suspect something is off in my spec file. Can you take a look? I can't seem to get the exact error (The output mentions checking build.log).

File: katello/katello.spec
Comment (mbacovsky) [inline katello/katello.spec]: It seems you need to prefix these with `%{buildroot}` see the build.log http://koji.katello.org/kojifiles/work/tasks/9917/29917/build.log

Comment (johnpmitsch) [reply katello/katello.spec]: Thanks! I updated

**Comment by mbacovsky** [comment]
The build is green now, the rpm can be downloaded from the task [1] for easier testing.
[1] http://koji.katello.org/koji/taskinfo?taskID=29919

**Comment by jturel** [comment]
Everything is still good with the change to the KatelloUtilities namespace. 👍

**Comment by johnpmitsch** [comment]
@jturel @evgeni Can you double check that all the changes that have been made since this PR has been open are correctly included in this PR? I wasn't able to resolve any merge conflicts since the files moved and changed so drastically, so applied them manually.

https://github.com/Katello/katello-packaging/commits/master/katello/katello-change-hostname

**Comment by johnpmitsch** [comment]
Updating to use accept an argument for the accepted_scenarios (again, for backporting). Changes are in a separate commit

This will allow us to overwrite the `accepted_scenarios` method, as shown below:
```
module Foo
  def scenarios
    @accepted || ["in module"]
  end
end

class Bar
  include Foo

  def initialize(accepted=nil)
    @accepted = accepted
  end
end
```
```
2.2.4 :002 > bar_no_params = Bar.new
 => #<Bar:0x00000000b1dcd0 @accepted=nil> 
2.2.4 :003 > bar_no_params.scenarios
 => ["in module"] 
2.2.4 :004 > bar_with_params = Bar.new(["not in module"])
 => #<Bar:0x00000000a8cb90 @accepted=["not in module"]> 
2.2.4 :005 > bar_with_params.scenarios
 => ["not in module"] 
```

**Comment by johnpmitsch** [comment]
Testing out the accepted scenarios changing:
```diff
diff --git a/katello/katello-backup b/katello/katello-backup
index 745143e..f74b3c8 100755
--- a/katello/katello-backup
+++ b/katello/katello-backup
@@ -1,7 +1,7 @@
 #!/usr/bin/env ruby
 require "/usr/share/katello/backup.rb"
 
-katello_backup = KatelloUtilities::Backup.new("foreman-proxy-content", "foreman proxy")
+katello_backup = KatelloUtilities::Backup.new("foreman-proxy-content", "foreman proxy", ["blah"])
 
 if katello_backup.accepted_scenarios.include? katello_backup.last_scenario
   katello_backup.run
diff --git a/katello/katello-change-hostname b/katello/katello-change-hostname
index b5029c0..e4d5f8d 100755
--- a/katello/katello-change-hostname
+++ b/katello/katello-change-hostname
@@ -1,7 +1,7 @@
 #!/usr/bin/env ruby
 require "/usr/share/katello/hostname-change.rb"
 
-hostname_change = KatelloUtilities::HostnameChange.new("Foreman Proxy", "Foreman Proxies", "foreman-proxy-content")
+hostname_change = KatelloUtilities::HostnameChange.new("Foreman Proxy", "Foreman Proxies", "foreman-proxy-content", nil, nil, ["blah"])
 
 if hostname_change.accepted_scenarios.include? hostname_change.last_scenario
   hostname_change.run
diff --git a/katello/katello-restore b/katello/katello-restore
index e0f7bca..c3fce8f 100755
--- a/katello/katello-restore
+++ b/katello/katello-restore
@@ -1,7 +1,7 @@
 #!/usr/bin/env ruby
 require "/usr/share/katello/restore.rb"
 
-katello_restore = KatelloUtilities::Restore.new("foreman-proxy-content")
+katello_restore = KatelloUtilities::Restore.new("foreman-proxy-content", ["blah"])
 
 if katello_restore.accepted_scenarios.include? katello_restore.last_scenario
   katello_restore.run
```
```
[root@change-hostname-34 katello]# ./katello-backup /backup
This utility can't run on a non-katello system.
[root@change-hostname-34 katello]# ./katello-restore /backup
This utility can't run on a non-katello system.
[root@change-hostname-34 katello]# ./katello-change-hostname -u admin -p changeme -d foo.example.com
This utility can't run on a non-katello system.
```
An error message is displayed when you pass in a parameter that is not an accepted scenario to the classes

**Comment by jturel** [comment]
@johnpmitsch All the porting of recent changes (from 9-14 to present) look good!

**Comment by johnpmitsch** [comment]
@ehelms can you look over the recent changes?

File: katello/backup.rb
Comment (evgeni) [inline katello/backup.rb]: This makes it impossible to test the code from a git checkout :(

How about

```ruby
require File.join(File.expand_path('..', __FILE__), 'helper.rb')
```

instead? :)

File: katello/backup.rb
Comment (evgeni) [inline katello/backup.rb]: Should this be moved to `helper.rb`?

File: katello/hostname-change.rb
Comment (evgeni) [inline katello/hostname-change.rb]: Should this be moved to `helper.rb`?

File: katello/hostname-change.rb
Comment (evgeni) [inline katello/hostname-change.rb]: Should this be moved to `helper.rb`?

File: katello/hostname-change.rb
Comment (evgeni) [inline katello/hostname-change.rb]: Should this be moved to `helper.rb`?

File: katello/hostname-change.rb
Comment (evgeni) [inline katello/hostname-change.rb]: Should this be moved to `helper.rb`?

File: katello/restore.rb
Comment (evgeni) [inline katello/restore.rb]: This makes it impossible to test the code from a git checkout :(

File: katello/katello-restore
Comment (evgeni) [inline katello/katello-restore]: This makes it impossible to test the code from a git checkout :(

File: katello/katello-backup
Comment (evgeni) [inline katello/katello-backup]: This makes it impossible to test the code from a git checkout :(

File: katello/katello-change-hostname
Comment (evgeni) [inline katello/katello-change-hostname]: This makes it impossible to test the code from a git checkout :(

File: katello/hostname-change.rb
Comment (evgeni) [inline katello/hostname-change.rb]: This makes it impossible to test the code from a git checkout :(

File: katello/katello.spec
Comment (evgeni) [inline katello/katello.spec]: at this point I wonder if we should create a proper katello_utils gem, put all the real code there, and just package require that?

**Review by evgeni — CHANGES_REQUESTED** [review]
after reading this, I think we should go a step further and make the code a real gem, not a random collection of .rb files that live in katello-packaging.

I did not verify if you can or cannot do copy&paste properly ;)

File: katello/katello.spec
Comment (mbacovsky) [reply katello/katello.spec]: @evgeni I thought there was general agreement on moving the utils to `foreman-maintain`in a near future. Do you think it makes sense to invest time in creating a temporary gem?

Comment (evgeni) [reply katello/katello.spec]: Right, I remember hearing that. I guess it highly depends on the time-schedule then?
Creating a gem is (IMHO) pretty easy and cheap, unless we're moving to foreman-maintain in 2 wks or something.
The code requires a class-ification for either path (pure class-ification, a k_u gem or f_m gem) and if make a gem now, the move to foreman-maintain is just a rename of the files away.

File: katello/backup.rb
Comment (johnpmitsch) [reply katello/backup.rb]: @evgeni I put [instructions](https://git.io/vdklj) here a few times on how to check this out for testing. Those will be a little out of date, but I also tried to update it in newer comments (that may be hidden by github). 

Will the above solution place the files in `/usr/share/katello` in a production install? That is where we agreed to put the utility class files

Comment (evgeni) [reply katello/backup.rb]: The above does not move the files anywhere, it just requires them with a relative path.
We even could just use require_relative here in most cases.

**Comment by johnpmitsch** [comment]
@evgeni This is meant to be a "step in the right direction" change to start the utilities moving over to foreman-maintain. I am not sure I would want to change this to be a gem before doing that, I plan to start on the effort after this gets merged.

I agree on the further DRYing up of the classes, but wanted to keep things simple in this first refactoring and then follow up with smaller additional changes (moving other scripts into classes and removing duplicate commands). I would rather merge this first so any additional changes anyone makes to the three refactored scripts can use this structure and I don't have to keep manually applying changes to master to this PR. Of course, as always, I'm open to discussion :)

File: katello/backup.rb
Comment (johnpmitsch) [reply katello/backup.rb]: @evgeni how does that work with the rpm install?

File: katello/katello.spec
Comment (mbacovsky) [reply katello/katello.spec]: @evgeni I see your point and it makes sense. However I don't expect the move to be just a rename. `foreman-maintain` is supposed to work on all the scenarios we have (foreman, katello, various combination of the plugins, upstream, downstream). f-m as a framework introduces `features`and behaves on their presence or absence. Some of the logic is already there some will need to be extended. I expect the code here to get restructured to match f-m's layout. With that in mind I just wanted to rise discussion if further classification still makes sense.

**Comment by johnpmitsch** [comment]
@evgeni I updated according to your suggestions, can you give it a second look?

**Comment by ehelms** [comment]
@johnpmitsch I merged your man-page PR and I think I broke the spec file. Please rebase.

**Comment by ehelms** [comment]
@evgeni As stated, the general agreement and discussion was to do this a as a step towards foreman_maintain, without going full scale in creating a new gem, new packaging, new everything. While this will likely persist a full release, I don't agree with breaking it out at this time. Yes, it makes this place uglier, but the intent was to do as little as possible to re-factor into re-usable classes to facilitate easier re-use and a shortened timescale. Another way to look at it is, we want to fix the "bug" not turn this into a feature.

I should also note that I agree with @johnpmitsch in that, in this case, re-factoring here will lead to higher risk rather than lower risk due to the amount and complexity of changes that are introduced. By first wrapping existing functionality into classes with as few new methods as possible we retain the original behaviors enough to regression test easier. The more that changes, the more edge cases we potentially introduce on untested code.

The ideal move to foreman_maintain will introduce rubocop, and testing and a singular framework for use which will provide a lot more robustness.

**Comment by ehelms** [comment]
Looks like yet again a round of re-factors has occurred meaning this will need the full gambit of re-testing. Does that fall to you @jturel ?

**Comment by johnpmitsch** [comment]
@ehelms thanks, I rebased the man page, please double check it looks ok. @evgeni is going to take another look at the latest refactor

@jturel I would recommend holding off on testing until @evgeni takes a look and comments, that way we can save you some unnecessary testing if there are additional changes :)

File: katello/restore.rb
Comment (evgeni) [inline katello/restore.rb]: this has to be `@optparse`, not `opts`

File: katello/restore.rb
Comment (evgeni) [inline katello/restore.rb]: this has to be `@optparse`, not `opts`

File: katello/restore.rb
Comment (evgeni) [inline katello/restore.rb]: this has to be `@optparse`, not `opts`

**Review by evgeni — COMMENTED** [review]
This is clearly a step in the right direction! I think going without the gem until we merge into foreman-maintain is OK - esp as I missed the "we need to do the packaging for the gem" part in my initial "it's soo easy" comment.

There is one comment on restore.rb. ACK as soon as that has been adressed!

:sheep:

**Comment by johnpmitsch** [comment]
@evgeni nice catch! I updated @jturel lets test this out again (hopefully for the last time)

**Review by ehelms — APPROVED** [review]

**Comment by jturel** [comment]
I was able to verify all scripts against a 3.4 installation and a proxy!

**Pull request merged**
